<a href="https://colab.research.google.com/github/elterco83/Coder_DS1/blob/main/Dataset_Etchezar_Final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **DATA SCIENCE I / ENTREGA FINAL**

# Dataset : Ventas Adidas USA

**Abstract**


Este dataset proporciona una cantidad significativa de datos sobre ventas, lo que permitirá realizar un análisis exhaustivo y obtener conclusiones precisas sobre el rendimiento comercial de Adidas en el mercado estadounidense.

El objetivo de este proyecto es analizar las ventas de la marca en los Estados Unidos entre enero de 2020 y febrero de 2021, con el fin de identificar tendencias, patrones de comportamiento del consumidor y oportunidades de optimización en la estrategia comercial.

El dataset incluye casi 10.000 registros de ventas individuales, detallando información clave sobre cada transacción. Entre las variables presentes se encuentran el método de venta (online, tiendas físicas, outlet), la ubicación geográfica (región, estado, ciudad), los productos comercializados y los retailers que venden la marca.

Además, se incluyen variables numéricas como el precio por unidad, las unidades vendidas, margen operativo, ganancia y el monto total de la venta, lo que permite un análisis detallado del impacto de cada factor en el desempeño comercial de Adidas.

La amplitud de los datos disponibles brinda la posibilidad de realizar análisis comparativos entre distintos canales de venta, evaluar el impacto de la ubicación en el volumen de ventas y detectar posibles fluctuaciones en la demanda según el tipo de producto y la temporada del año.

A partir de esta información, se podrá desarrollar un modelo predictivo que ayude a optimizar la toma de decisiones estratégicas para mejorar la eficiencia comercial y maximizar los ingresos de la empresa.

Procederemos a realizar la carga del archivo, como asi tabien la importacion de librerias

In [151]:
!wget https://raw.githubusercontent.com/elterco83/Coder_DS1/refs/heads/main/Ventas_Adidas_USA_Entrega_Final.csv
import pandas as pd
import numpy as np
import statsmodels.api as sm
df_base = pd.read_csv('Ventas_Adidas_USA_Entrega_Final.csv')

--2025-02-22 13:27:14--  https://raw.githubusercontent.com/elterco83/Coder_DS1/refs/heads/main/Ventas_Adidas_USA_Entrega_Final.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1301383 (1.2M) [text/plain]
Saving to: ‘Ventas_Adidas_USA_Entrega_Final.csv.9’

Ventas_Adidas_USA_E 100%[===================>]   1.24M  --.-KB/s    in 0.04s   

2025-02-22 13:27:14 (31.5 MB/s) - ‘Ventas_Adidas_USA_Entrega_Final.csv.9’ saved [1301383/1301383]



# Resumen de Atributos y sus descripciones



1.   ***ID_Ventas :*** Estas columna la usaremos como indice, y es el codigo identificador de cada venta.
2. ***fecha :*** El dia en cual se realizo la venta.
3. ***temporada:*** Estacion de año correspondiente a la fecha.

1.   ***sexo :*** La condicion sexual de cada cliente.
2.   ***retailer :*** Son los puntos de Venta, que intermedian entre la marca y el cliente.

1. ***region :*** Esta columna describe en que Region de USA se realizo la venta.

1.  ***estado :*** Esta columna describe en que Estado de la Region se realizo la venta.
2.  ***ciudad :*** Esta columna describe en que Ciudad de ese Estado se realizo la venta.

8.   ***metodo de venta :*** Hace referencia al formato en el cual se dio la venta, el cual es elegido por el cliente.

9.   ***producto :*** Lo ofrecido por la marca a sus clientes
2.   ***precio por unidad :*** El precio unitario de cada producto.

2.    ***unidades vendidas :*** La cantidad de unidades que tiene cada venta.

1.    ***ventas totales :*** Es el valor total de la factura.
2.   ***ganancia :*** La ganancia monetaria obtenida por la venta.

1.   ***margen operativo :*** Expresa el porcentaje que hay entre ganancia y la venta.





















**Análisis Exploratorio de los Datos**

A traves de conjunto de funciones de la libreria pandas para validaremos la calidad y estructura del DataFrame:


* **head:** Nos permite observar las primeras filas del DataFrame.

*  **info:** Resume registros, columnas, tipos y valores nulos.

* **describe:** Proporciona estadísticas descriptivas de las variables numéricas.
* **dtypes:** Verifica el tipo de dato de cada columna.

* **len:** Muestra el número total de registros.


Estas técnicas nos aseguran que los datos están limpios y listos para el análisis posterior.

In [152]:
df_base = df_base.reset_index(drop=True)
df_base.set_index('ID_Ventas', inplace=True)
df_base.head(20)

fecha temporada        sexo     retailer     region  \
ID_Ventas                                                                
FC-000000001  2020-01-01  Invierno        male  Foot Locker  Northeast   
FC-000000002  2020-01-02  Invierno        male  Foot Locker  Northeast   
FC-000000003  2020-01-03  Invierno  non-binary  Foot Locker  Northeast   
FC-000000004  2020-01-04  Invierno      female  Foot Locker  Northeast   
FC-000000005  2020-01-05  Invierno        male  Foot Locker  Northeast   
FC-000000006  2020-01-06  Invierno      female  Foot Locker  Northeast   
FC-000000007  2020-01-07  Invierno      female  Foot Locker  Northeast   
FC-000000008  2020-01-08  Invierno      female  Foot Locker  Northeast   
FC-000000009  2020-01-21  Invierno      female  Foot Locker  Northeast   
FC-000000010  2020-01-22  Invierno      female  Foot Locker  Northeast   
FC-000000011  2020-01-23  Invierno      female  Foot Locker  Northeast   
FC-000000012  2020-01-24  Invierno      female  Foot Locker  Northeast   
FC-000000013  2020-01-25  Invierno  non-binary  Foot Locker  Northeast   
FC-000000014  2020-01-26  Invierno  non-binary  Foot Locker  Northeast   
FC-000000015  2020-01-27  Invierno        male  Foot Locker  Northeast   
FC-000000016  2020-01-28  Invierno      female  Foot Locker  Northeast   
FC-000000017  2020-01-29  Invierno      female  Foot Locker  Northeast   
FC-000000018  2020-01-30  Invierno      female  Foot Locker  Northeast   
FC-000000019  2020-01-31  Invierno      female  Foot Locker  Northeast   
FC-000000020  2020-02-01  Invierno      female  Foot Locker  Northeast   

                estado    ciudad metodo de venta           producto  \
ID_Ventas                                                             
FC-000000001  New York  New York        In-store     Calzado Urbano   
FC-000000002  New York  New York        In-store  Calzado Deportivo   
FC-000000003  New York  New York        In-store     Calzado Urbano   
FC-000000004  New York  New York        In-store  Calzado Deportivo   
FC-000000005  New York  New York        In-store       Indumentaria   
FC-000000006  New York  New York        In-store       Indumentaria   
FC-000000007  New York  New York        In-store     Calzado Urbano   
FC-000000008  New York  New York          Outlet  Calzado Deportivo   
FC-000000009  New York  New York          Outlet     Calzado Urbano   
FC-000000010  New York  New York          Outlet  Calzado Deportivo   
FC-000000011  New York  New York          Outlet       Indumentaria   
FC-000000012  New York  New York          Outlet       Indumentaria   
FC-000000013  New York  New York          Outlet     Calzado Urbano   
FC-000000014  New York  New York          Outlet  Calzado Deportivo   
FC-000000015  New York  New York          Outlet     Calzado Urbano   
FC-000000016  New York  New York          Outlet  Calzado Deportivo   
FC-000000017  New York  New York          Outlet       Indumentaria   
FC-000000018  New York  New York          Outlet       Indumentaria   
FC-000000019  New York  New York          Outlet     Calzado Urbano   
FC-000000020  New York  New York          Outlet  Calzado Deportivo   

              precio por unidad  unidades vendidas  ventas totales  ganancia  \
ID_Ventas                                                                      
FC-000000001               50.0               1200         60000.0  30000.00   
FC-000000002               50.0               1000         50000.0  15000.00   
FC-000000003               40.0               1000         40000.0  14000.00   
FC-000000004               45.0                850         38250.0  13387.50   
FC-000000005               60.0                900         54000.0  16200.00   
FC-000000006               50.0               1000         50000.0  12500.00   
FC-000000007               50.0               1250         62500.0  31250.00   
FC-000000008               50.0                900         45000.0  13500.00   
FC-000000009               

In [153]:
df_base.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9648 entries, FC-000000001 to FC-000009648
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   fecha              9648 non-null   object 
 1   temporada          9648 non-null   object 
 2   sexo               9648 non-null   object 
 3   retailer           9648 non-null   object 
 4   region             9648 non-null   object 
 5   estado             9648 non-null   object 
 6   ciudad             9648 non-null   object 
 7   metodo de venta    9648 non-null   object 
 8   producto           9648 non-null   object 
 9   precio por unidad  9648 non-null   float64
 10  unidades vendidas  9648 non-null   int64  
 11  ventas totales     9648 non-null   float64
 12  ganancia           9648 non-null   float64
 13  margen operativo   9648 non-null   float64
dtypes: float64(4), int64(1), object(9)
memory usage: 1.1+ MB


In [154]:
df_base.describe().round(1)

precio por unidad  unidades vendidas  ventas totales  ganancia  \
count             9648.0             9648.0          9648.0    9648.0   
mean                45.2              256.9          9327.3    3442.5   
std                 14.7              214.3         14191.6    5419.3   
min                  7.0                0.0             0.0       0.0   
25%                 35.0              106.0           425.4     192.2   
50%                 45.0              176.0           957.6     437.1   
75%                 55.0              350.0         15000.0    5206.2   
max                110.0             1275.0         82500.0   39000.0   

       margen operativo  
count            9648.0  
mean                0.4  
std                 0.1  
min                 0.1  
25%                 0.4  
50%                 0.4  
75%                 0.5  
max                 0.8

In [155]:
df_base.dtypes

fecha                 object
temporada             object
sexo                  object
retailer              object
region                object
estado                object
ciudad                object
metodo de venta       object
producto              object
precio por unidad    float64
unidades vendidas      int64
ventas totales       float64
ganancia             float64
margen operativo     float64
dtype: object

In [156]:
len(df_base)

9648

# Modelo de Predicción de Ventas

---

El objetivo de este modelo es predecir las ventas totales de productos Adidas.

Antes de entrenarlo, se elimina la variable *'precio por unidad'*, para evitar que el modelo tenga información redundante y prevenir problemas de colinealidad.
Si la incluyera, el modelo podría depender demasiado de esta relación matemática en lugar de identificar patrones reales en los datos.

Después de ese ajuste, uso el método stepwise para seleccionar automáticamente las variables más importantes para la predicción, descartando las que no aportan mucho.

La técnica que aplicada es Regresión Lineal Múltiple, permitiendo entender cómo influyen distintos factores en las ventas.

La idea es que el modelo no solo haga buenas predicciones, sino que también ayude a tomar mejores decisiones basadas en datos.



**Proceso y Modelado de Datos para Prediccion de Ventas : Paso a Paso**

**1.**   En primer lugar convertimos las variables categoricas a numericas (dummies) , para poder utilizarlas en el modelo de Regresión Lineal Múltiple.

In [157]:
#1

#Se convierten las variables categóricas a dummies
categoricas = ['ciudad','estado','sexo', 'retailer', 'region', 'metodo de venta', 'producto']
df_base = pd.get_dummies(df_base, columns=categoricas, dtype=int)

**2.**   En este codigo, se modifica el tipo de dato de las columna 'fecha', para poder crear la columna 'temporada' desde los meses. Luego esa columna tambien la pasamos a dummies (numerica). Por ultimo se elimina la columna fecha, ya que no es necesaria para el modelo. Y se suigue el proceso con df_clean.

In [158]:
#2

# se convierte la columna 'fecha' a datetime
df_base['fecha'] = pd.to_datetime(df_base['fecha'], errors='coerce')

# se determina la temporada a partir del mes
def obtener_temporada(fecha):
    mes = fecha.month
    if mes in [12, 1, 2]:
        return 'Invierno'
    elif mes in [3, 4, 5]:
        return 'Primavera'
    elif mes in [6, 7, 8]:
        return 'Verano'
    elif mes in [9, 10, 11]:
        return 'Otoño'

# Creamos una nueva columna 'temporada' extrayendo la estación de la fecha
df_base['temporada'] = df_base['fecha'].apply(lambda x: obtener_temporada(x) if pd.notnull(x) else None)

# Convertir la variable 'temporada' a dummies y le eliminamos 1 categoria para evitar la multicolinealidad
df_base = pd.get_dummies(df_base, columns=['temporada'], drop_first=True)

# Eliminamos la columna fecha ya que no vamos a necesitarla para el modelo
df_clean = df_base.drop(columns=['fecha'])


**3.**   Calculo de metricas estadisticas previo a eliminar los Outliers del dataset, eliminacion de los mismos. Y luego ya sin valores extremos,  se vuelven a calcular la media, mediana y moda.
Se crea 'df_final', como el Dataframe preparado para el modelo.

In [176]:
#3

# Calcular métricas estadísticas sobre df_base
media = df_base.mean()        # Media
mediana = df_base.median()    # Mediana
moda = df_base.mode().iloc[0] # Moda (toma la primera si hay múltiples)

# Mostrar resultados
print("Media con outliers:\n", media, "\n")
print("Mediana con outliers:\n", mediana, "\n")
print("Moda con outliers :\n", moda, "\n")
print("---------" * 8)



# Calculamos los cuartiles y el IQR
Q1 = df_clean['ventas totales'].quantile(0.25)
Q3 = df_clean['ventas totales'].quantile(0.75)
IQR = Q3 - Q1

# Definimos los límites inferior y superior
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Eliminamos lo outliers de df_clean y creamos df_final para usar en el modelo
df_final = df_clean[(df_clean['ventas totales'] >= lower_bound) & (df_clean['ventas totales'] <= upper_bound)]

print("Tamaño original:", df_base.shape)
print("Tamaño después de eliminar outliers:", df_final.shape)
print("---------" * 8)

# Calcular métricas estadísticas sobre df_final
media = df_final.mean()        # Media
mediana = df_final.median()    # Mediana
moda = df_final.mode().iloc[0] # Moda (toma la primera si hay múltiples)

# Mostrar resultados
print("Media sin outliers :\n", media, "\n")
print("Mediana sin outliers :\n", mediana, "\n")
print("Moda sin outliers :\n", moda, "\n")



Media con outliers:
 fecha                      2021-05-10 15:20:44.776119296
precio por unidad                              45.216625
unidades vendidas                             256.930037
ventas totales                                9327.34375
ganancia                                     3442.524476
                                       ...              
producto_Calzado Urbano                         0.333541
producto_Indumentaria                           0.333126
temporada_Otoño                                 0.252177
temporada_Primavera                             0.255286
temporada_Verano                                0.244196
Length: 131, dtype: object 

Mediana con outliers:
 fecha                      2021-06-04 00:00:00
precio por unidad                         45.0
unidades vendidas                        176.0
ventas totales                           957.6
ganancia                               437.142
                                  ...         
producto_Calzado U

**4.** En este paso, definimos como variables predictoras (X) a todas, excepto a la variable objetivo (y).

In [160]:
#4

X = df_final.drop(columns=['ventas totales']) # variables predictorias
y = df_final['ventas totales'] # variable objetivo


**5.**  Luego se fuerza la conversion de todas las variables a float64 para evitar errores en el modelo y reinicio los índices para mantener alineados X e y, asegurando un dataset limpio y listo para el entrenamiento

In [161]:
#5

# Imprime los types actuales para diagnóstico
print("Tipos de datos en X antes de forzar conversión:")
print(X.dtypes)
print("Tipo de y antes de forzar conversión:", y.dtype)

# Forzar la conversión
X = X.apply(lambda col: pd.to_numeric(col, errors='coerce')).astype(float)
y = pd.to_numeric(y, errors='coerce').astype(float)

# Reiniciar el índice para evitar que el índice cause problemas
X = X.reset_index(drop=True)
y = y.reset_index(drop=True)

# Verifica nuevamente
print("Tipos de datos en X después de la conversión:")
print(X.dtypes)
print("Tipo de y después de la conversión:", y.dtype)


Tipos de datos en X antes de forzar conversión:
precio por unidad          float64
unidades vendidas            int64
ganancia                   float64
margen operativo           float64
ciudad_Albany                int64
                            ...   
producto_Calzado Urbano      int64
producto_Indumentaria        int64
temporada_Otoño               bool
temporada_Primavera           bool
temporada_Verano              bool
Length: 129, dtype: object
Tipo de y antes de forzar conversión: float64
Tipos de datos en X después de la conversión:
precio por unidad          float64
unidades vendidas          float64
ganancia                   float64
margen operativo           float64
ciudad_Albany              float64
                            ...   
producto_Calzado Urbano    float64
producto_Indumentaria      float64
temporada_Otoño            float64
temporada_Primavera        float64
temporada_Verano           float64
Length: 129, dtype: object
Tipo de y después de la conversión: 

**6.** Se aplica la selección de variables stepwise , las misma prepara un sub-conjunto de predictores en función de su p-valor.
Asegurando un modelo más eficiente con solo las variables más importantes.

In [162]:
#6

def stepwise_selection(X, y, initial_list=[], threshold_in=0.01, threshold_out=0.05, verbose=True):
    """
    Selección stepwise para encontrar el subconjunto óptimo de predictores.

    Parámetros:
      X - DataFrame con las variables predictoras.
      y - Serie con la variable objetivo.
      initial_list - Lista de variables a incluir inicialmente.
      threshold_in - Umbral para agregar una variable.
      threshold_out - Umbral para eliminar una variable.
      verbose - Si True, imprime cada paso.

    Retorna:
      Lista de variables seleccionadas.
    """
    included = list(initial_list)
    while True:
        changed = False

        # Paso forward: intentar agregar variables
        excluded = list(set(X.columns) - set(included))
        new_pval = pd.Series(index=excluded, dtype=float)
        for new_column in excluded:
            model = sm.OLS(y, sm.add_constant(X[included + [new_column]])).fit()
            new_pval[new_column] = model.pvalues[new_column]
        best_pval = new_pval.min() if not new_pval.empty else None
        if best_pval is not None and best_pval < threshold_in:
            best_feature = new_pval.idxmin()
            included.append(best_feature)
            changed = True
            if verbose:
                print('Agregamos {:30} con p-valor {:.6}'.format(best_feature, best_pval))

        # Paso backward: intentar eliminar variables
        model = sm.OLS(y, sm.add_constant(X[included])).fit()
        # Se excluye el intercepto
        pvalues = model.pvalues.iloc[1:]
        worst_pval = pvalues.max() if not pvalues.empty else None
        if worst_pval is not None and worst_pval > threshold_out:
            worst_feature = pvalues.idxmax()
            included.remove(worst_feature)
            changed = True
            if verbose:
                print('Eliminamos {:30} con p-valor {:.6}'.format(worst_feature, worst_pval))

        if not changed:
            break

    return included



**7.**   Antes de ejecutar el modelo stepwise, se elimina la variable *'precio por unidad'*, para evitar que el modelo tenga información redundante y prevenir problemas de colinealidad.

In [163]:
#7

#se elimina una variable predictoria de X antes de las seleccion
X = X.drop(columns=['precio por unidad'], errors='ignore')


**8.**  Se ejecuta stepwise_selection sobre X e y, dando como resultado las variables seleccionadas para ejecutar el modelo de regresion lineal multiple.

In [177]:
#8

# Ejecutamos la selección de variables stepwise
selected_features = stepwise_selection(X, y, verbose=True)

print("Variables seleccionadas:", selected_features)



Agregamos metodo de venta_In-store       con p-valor 0.0
Agregamos ganancia                       con p-valor 0.0
Agregamos margen operativo               con p-valor 0.0
Agregamos unidades vendidas              con p-valor 6.11249e-173
Agregamos metodo de venta_Outlet         con p-valor 1.15148e-75
Agregamos metodo de venta_Online         con p-valor 2.32699e-286
Agregamos ciudad_New York                con p-valor 1.16559e-24
Agregamos ciudad_San Francisco           con p-valor 4.1275e-20
Agregamos estado_Hawaii                  con p-valor 1.93774e-17
Agregamos ciudad_Honolulu                con p-valor 1.93774e-17
Agregamos producto_Indumentaria          con p-valor 4.2469e-15
Agregamos estado_Alaska                  con p-valor 5.13768e-14
Agregamos ciudad_Anchorage               con p-valor 5.13768e-14
Agregamos estado_Washington              con p-valor 1.13522e-10
Agregamos ciudad_Seattle                 con p-valor 1.13522e-10
Agregamos region_Northeast               con p-va

**9.**   Para la ejecucion de codigo, almacenamos X (variables seleccionadas) e y (variable objetivo) código, luego se dividen los datos para entrenar y probar al modelo y ya con esa informacion pasamos a  entrenar un modelo de regresión lineal.

In [178]:
#9

#importacion de librerias para el entrenamiento del modelo

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# preparamos X e y para el modelo
X_final = df_final[selected_features]
y_final = df_final['ventas totales']

# Dividimos X para entrenar y probar implementando la semilla
X_train, X_test, y_train, y_test = train_test_split(X_final, y_final, test_size=0.2, random_state=42)

# Modelo de Regresión Lineal (para comparar)
modelo_lineal = LinearRegression()
modelo_lineal.fit(X_train, y_train)
y_pred_lineal = modelo_lineal.predict(X_test)




**10.** El ultimo paso, evaluamos el modelo con métricas como MSE, RMSE y R².
Tambien calculasmos que porcetaje representa el RSME en el promedio de ventas totales.

In [189]:
#10

# Importacion de libreria para la Evaluacion del modelo
from sklearn.metrics import mean_squared_error, r2_score

mse_lineal = mean_squared_error(y_test, y_pred_lineal)
rmse_lineal = np.sqrt(mse_lineal)
r2_lineal = r2_score(y_test, y_pred_lineal)

print("Modelo de Regresion Lineal:")
print(f"MSE en dolares al cuadrado: {mse_lineal} USD²")
print(f" RMSE en dolares: {rmse_lineal} USD")
print("R²:" ,r2_lineal)

#calculo del promedio de ventas
promedio_ventas = y.mean()
print("Promedio de ventas (en USD ):", promedio_ventas / 1000)

#porcentaje de rsme sobre el pomedio de ventas
rmse_lineal = np.sqrt(mse_lineal)
error_relativo = (rmse_lineal / promedio_ventas) * 100
print("Porcentaje de error relativo:", error_relativo, "%")


Modelo de Regresion Lineal:
MSE en dolares al cuadrado: 3152437.1796400966 USD²
 RMSE en dolares: 1775.5103997555454 USD
R²: 0.9664851756532985
Promedio de ventas (en USD ): 6.507841856586992
Porcentaje de error relativo: 27.28262977008946 %


 **CONCLUSION FINAL**

En este análisis, utilicé un modelo de Regresión Lineal Múltiple para predecir las ventas totales, aplicando un enfoque de selección de variables para mejorar la precisión del modelo.

Antes de la selección de variables, decidí eliminar la variable "precio por unidad", ya que esta se encontraba implícita en el cálculo de ventas totales (precio por unidad × unidades vendidas). Mantener esta variable habría significado darle al modelo la respuesta de manera directa, lo que afectaría su capacidad de identificar patrones reales.

Luego, se aplicó el método Stepwise Selection, que permitió seleccionar únicamente aquellas variables que tienen un impacto significativo en las ventas, eliminando las que no aportaban valor al modelo.

Como resultado, se obtuvo un modelo con un coeficiente de determinación (R²) de 0.966, lo que indica que las variables seleccionadas explican el 96.6% de la variabilidad de las ventas.

Sin embargo, el error cuadrático medio (MSE) de 3,152,437 y el RMSE de 1,775.51 reflejan que el modelo aún tiene diferencias significativas entre las predicciones y los valores reales.

Esto se traduce en un error relativo del 27.28%, lo que sugiere que en algunos casos la diferencia entre lo estimado y lo observado es considerable.